embassy-time/src/queue.rs

```diff
@@ -0,0 +1,617 @@
+//! Generic timer queue implementation
+//!
+//! This module provides a timer queue that works with any executor.
+//!
+//! In terms of performance, this queue will likely be less efficient in comparison to executor-native queues,
+//! like the one provided with e.g. the `embassy-executor` crate.
+//!
+//! # Enabling the queue
+//! - Enable the Cargo feature `generic-queue`. This will automatically instantiate the queue.
+//!
+//! # Initializing the queue
+//! - Call ```unsafe { embassy_time::queue::initialize(); }``` early on in your program, before any of your futures that utilize `embassy-time` are polled.
+//!
+//! # Customizing the queue
+//! - It is possible to customize two aspects of the queue:
+//!   - Queue size:
```

Comment:
Maybe queue size could be configured via Cargo features? I don't like that there's 2 ways of initializing the queue, either implicitly or via the macro. This'd allow doing it always implicitly.

Reply:
With the circular dependencies that are being introduced by using `extern "Rust"` magic, the *implicits might not always work* (that's the topic we discussed on the forum last week). I'll elaborate more on that in a separate yet-to-be-opened-issue, but for now I'm relatively convinced that you are not hitting the linker issue only because you don't have yet a more elaborate crate dependency graph that I do. But the consequences are that - unless we want to change the whole `extern "Rust"` magic approach with something else (does this "else" even exist, I wonder) - in some cases user* should be calling the following macros *explicitly from their binary crate* so as not to run into linking issues:
- `critical_section::set_impl!`
- `embassy_time::driver_impl!`
- Possibly `embassy_time::generic_queue!`

So if we are to expose these macros as a "backup" approach anyway, why not keep the configuration that way as well?